[rubocop] Step 5

AliSoftware opened the pull request:
Continuing work on #207 and rubocop.

⚠️ Some of the rules that have been incrementally enabled in this PR were not auto-corrected by rubocop but were corrected manually – so will need more attentive review.

This also includes a small refactor (consisting of the extraction of long portion of code into a separate method, to reduce the code complexity and method length, see https://github.com/wordpress-mobile/release-toolkit/pull/214/commits/f84285d778a7d770c7adcf9fc836784777ef705e) that might also need careful review – especially to ensure that all the variables used in the original code were all passed to the method and that the extracted code didn't modify any variable that would then later be used after it in the original code, etc.

The corresponding commits are labeled accordingly (`[rubocop] fix … (manual)` and the like) – in case you want to review commit-per-commit.

AliSoftware committed:
[rubocop] fix Style/AndOr

AliSoftware committed:
[rubocop] fix Style/BarePercentLiterals

AliSoftware committed:
[rubocop] fix Style/BlockDelimiters

AliSoftware committed:
[rubocop] fix Style/BracesAroundHashParameters

AliSoftware committed:
[rubocop] fix Style/ClassAndModuleChildren

AliSoftware committed:
[rubocop] fix Style/ClassCheck

AliSoftware committed:
[rubocop] fix Style/ConditionalAssignment

AliSoftware committed:
[rubocop] fix Style/DefWithParentheses

AliSoftware committed:
[rubocop] fix Style/EmptyLiteral

AliSoftware committed:
[rubocop] configure Style/EmptyMethod

AliSoftware committed:
[rubocop] fix Style/Encoding

AliSoftware committed:
[rubocop] fix Style/ExpandPathArguments

AliSoftware committed:
[rubocop] fix Style/FormatString

AliSoftware committed:
[rubocop] disable Style/FrozenStringLiteralComment

Because enabling it, even if it's recommended, require very careful tests and review to avoid accidental runtime crashes

AliSoftware committed:
[rubocop] fix Style/HashSyntax

AliSoftware committed:
[rubocop] fix Style/IfUnlessModifier

AliSoftware committed:
[rubocop] fix Style/InverseMethods

AliSoftware committed:
[rubocop] fix Style/ParenthesesAroundCondition

AliSoftware committed:
[rubocop] fix Style/MultilineIfThen

AliSoftware committed:
[rubocop] fix Style/NegatedIf

AliSoftware committed:
[rubocop] fix Style/NegatedUnless

AliSoftware committed:
[rubocop] fix Style/NilComparison

AliSoftware committed:
[rubocop] fix Style/NonNilCheck (safe-auto-correct)

AliSoftware committed:
[rubocop] fix Style/NonNilCheck (manual fixes) ⚠️

AliSoftware committed:
[rubocop] ⚠️ Manual fix of Metrics/AbcSize

Used RubyMine's refactoring feature to extract the big chunk of code building the entries into its private method, to reduce the self.run(…)'s complexity

AliSoftware committed:
[rubocop] fix Style/Not

AliSoftware committed:
[rubocop] fix Style/NumericLiteralPrefix

AliSoftware committed:
[rubocop] fix Style/PercentLiteralDelimiters

AliSoftware committed:
[rubocop] fix Style/RegexpLiteral

AliSoftware committed:
[rubocop] fix Style/Semicolon

AliSoftware committed:
[rubocop] fix Style/StringLiteralsInInterpolation

AliSoftware committed:
[rubocop] fix Style/TrailingCommaInArguments

AliSoftware committed:
[rubocop] fix Style/TrailingCommaInArrayLiteral

AliSoftware committed:
[rubocop] fix Style/TrailingCommaInHashLiteral

jkmassel commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/promo_screenshots_helper.rb:
```diff
@@ -410,14 +410,14 @@ def resolve_path(path)
 
       def resolve_text_into_path(text, locale)
-        localizedFile = sprintf(text, locale)
+        localizedFile = format(text, locale)
```
I had to look up `format`, but it looks like this is a mechanical transformation – functionally identical to `sprintf` in this context. 👍

AliSoftware replied:
Yep, I used `--safe-auto-correct` (as opposed to just `--auto-correct`) for that one and indeed checked too, they behave the same there (which the rule's autocorrect was smart enough to ensure apparently) 👍

jkmassel commented on .rubocop.yml:
```diff
@@ -19,6 +19,13 @@ Lint/UnusedMethodArgument:
+# Enabling the 'frozen_string_literal: true' comment is nice but should be done with extra care
+# because enabling it on code that accidentally mutates string literals will crash at runtime
+# (e.g. `name = 'Hello'` then later `name << ' world'` will crash if the comment is enabled)
+# So we might enable it at some point, be we will need very careful review that this doesn't break everything.
```
👍

jkmassel commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/an_metadata_update_helper.rb:
```diff
@@ -137,12 +137,16 @@ def initialize(block_key, content_file_path, release_version)
         super(fw) unless File.zero?(@content_file_path)
       end
+
+      def is_int?(value)
```
This looks ok to me – functionally identical to what came before.

The conditionals that reference this though IMHO could do with some cleanup and documentation though in the future 😅

AliSoftware replied:
Yeah, lots of ternary operator shenanigans I agree. I plan to tackle that probably during #203.

jkmassel commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/promo_screenshots_action.rb:
```diff
@@ -238,6 +173,75 @@ def self.available_options
         true
       end
+
+      def self.build_entries(config, languages, output_directory, params)
```
I like this, but AFAICT `config` isn't used directly and we could just pass the entries. 

Might not be worth fixing just now

AliSoftware replied:
Nice catch! Fixed in fd8d9bde9600caa7a283eb1e46600811c3a03d50

(That refactoring might go even further I guess, but didn't want to spend too much time on it this time around as I preffered to focus on move things forward in order to put rubocop stuff behind me ASAP. Might revisit this in the future)

jkmassel commented on spec/ios_lint_localizations_spec.rb:
```diff
@@ -37,7 +37,7 @@
           SCRIPT
           FileUtils.mkdir_p File.join(install_dir, 'bin')
-          File.write(File.join(install_dir, 'bin/swiftgen'), script, perm: 0766)
+          File.write(File.join(install_dir, 'bin/swiftgen'), script, perm: 0o766)
```
Is there a format that can do like `0x766` or something? 

Not a huge deal, but I feel like if I look at this in six months I'm gonna be like "what _is_ that string?"

But then I'd probably just look at the `File.write` call and figure it out in like 2 seconds, so not a huge deal – more just curious as it seems unusual.

AliSoftware replied:
> Is there a format that can do like 0x766 or something?

`0x` is the prefix for hexadecimal, while `0o` is the prefix for octal, which is what is used with `chmod`-like permissions. So `0o` is that format that can do it :P

Note: I looked at https://ruby-doc.org/core-2.5.0/File/Constants.html to see if there were constants we could bitwise-or to provide the same value in a more readable way, but those seems to be focused on providing constants for things like the open modes (read/read-write/append, text/binary) not on the Unix permissions.

Agree that since `0o` is not as widely used/known as `0x` it might need a quick comment though.

```suggestion
          # note: `0o` is octal notation, used to specify chmod-like flags
          File.write(File.join(install_dir, 'bin/swiftgen'), script, perm: 0o766)
```

jkmassel commented on .rubocop_todo.yml:
```diff
@@ -464,13 +464,6 @@ Style/TernaryParentheses:
-# Cop supports --auto-correct.
-# Configuration parameters: EnforcedStyleForMultiline.
-# SupportedStylesForMultiline: comma, consistent_comma, no_comma
-Style/TrailingCommaInArrayLiteral:
```
I'm surprised by this one because typically we'd leave it in case we needed to add another entry later.

I presume this was just the less-destructive approach?

AliSoftware replied:
It was. But good point, we could switch the rule the other way and require them all the time, I'm fine with it too, will see how much changes it would make to turn it on.

AliSoftware replied:
rule flipped in f6499f759a22dca36b72b136bca963c66b9f25cf, agree it's nicer that way

jkmassel reviewed (approved):
This LGTM – I left a few questions, but nothing blocking

AliSoftware committed:
Add comment about octal notation for chmod-like flags

AliSoftware committed:
[rubocop] fix Style/TrailingCommaInArrayLiteral

Change the setting of the rule from the default (no_comma) to consistent_comma, which better matches our style

AliSoftware committed:
build_entries doesn't need the whole config Hash

AliSoftware commented:
Thanks for the review @jkmassel 👍  I should have addressed all your questions and suggestions 🙂
